emigration.lua

```diff
@@ -39,45 +38,114 @@ function desireToStay(unit,method,civ_id)
 end
 
 function desert(u,method,civ)
-    u.following = nil
-    local line = dfhack.TranslateName(dfhack.units.getVisibleName(u)) .. " has "
-    if method == 'merchant' then
-        line = line.."joined the merchants"
-        u.flags1.merchant = true
-        u.civ_id = civ
-    elseif method == 'diplomat' then
-        line = line.."followed the diplomat"
-        u.flags1.diplomat = true
-        u.civ_id = civ
-    else
-        line = line.."abandoned the settlement in search of a better life."
-        u.civ_id = -1
-        u.flags1.forest = true
-        u.animal.leave_countdown = 2
-    end
-    print(line)
-    dfhack.gui.showAnnouncement(line, COLOR_WHITE)
+	u.following = nil
+	local line = dfhack.TranslateName(dfhack.units.getVisibleName(u)) .. " has "
+	if method == 'merchant' then
+		line = line.."joined the merchants"
+		u.flags1.merchant = true
+		u.civ_id = civ
+	else
+		line = line.."abandoned the settlement in search of a better life."
+		u.civ_id = civ
+		u.flags1.forest = true
+		u.flags2.visitor = true
+		u.animal.leave_countdown = 2
+	end
+
+	local hf_id = u.hist_figure_id
+	local hf = df.historical_figure.find(u.hist_figure_id)
+	local fort_ent = df.global.ui.main.fortress_entity
+	local civ_ent = df.historical_entity.find(hf.civ_id)
+
+	local newent_id = -1
+	local newsite_id = -1
+	
+	-- free owned rooms
+	for i = #u.owned_buildings-1, 0, -1 do
+		local temp_bld = df.building.find(u.owned_buildings[i].id)
+		dfhack.buildings.setOwner(temp_bld, nil)
+	end
+
+	-- erase the unit from the fortress entity
+	for k,v in pairs(fort_ent.histfig_ids) do
```

Comment:
while `pairs` likely still works, the semantics of `ipairs` are more appropriate. here and in all the loops below